### PR TITLE
Features/mimisbrunnr

### PR DESCRIPTION
**Pull request by hskim881028**
Added actions and sheets for  mimisbrunnr

### CODE REVIEW

File: Lib9c/TableData/MimisbrunnrSheet.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                    }
+                }
+            }
+        }
+        
+        public MimisbrunnrSheet() : base(nameof(MimisbrunnrSheet))
+        {
+        }
+    }
+}
```
Comment (ipdae) [inline Lib9c/TableData/MimisbrunnrSheet.cs:37]:
```suggestion
}

```

File: .Lib9c.Tests/Action/MimisbrunnrBttleTest.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            var value = nextState.GetState(_rankingMapAddress);
+
+            var rankingMapState = new RankingMapState((Dictionary)value);
+            var info = rankingMapState.GetRankingInfos(null).First();
+
+            Assert.Equal(info.AgentAddress, _agentAddress);
+            Assert.Equal(info.AvatarAddress, _avatarAddress);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (hskim881028) [reply .Lib9c.Tests/Action/MimisbrunnrBttleTest.cs:184]: I forgot eof again T.T

File: Lib9c/Action/MimisbrunnrBattle.cs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+using Nekoyume.TableData;
+using Serilog;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
+    [ActionType("mimisbrunnr_battle")]
+    public class MimisbrunnrBattle : GameAction
+    {
+        public List<int> costumes;
```
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Action/MimisbrunnrBattle.cs:22]: Equipments type is guid, but costumes type is integer

File: Lib9c/Action/InvalidElementalException.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+        public InvalidElementalException(string message) : base(message)
+        {
+        }
+
+        protected InvalidElementalException(SerializationInfo info, StreamingContext context) : base(info, context)
+        {
+        }
+    }
+}
```
Comment (ipdae) [inline Lib9c/Action/InvalidElementalException.cs:21]:
```suggestion
}

```

File: Lib9c/Action/MimisbrunnrBattle.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Action/MimisbrunnrBattle.cs:22]: I think it is better to fix after merging PR(https://github.com/planetarium/lib9c/pull/234/commits/7faf3b60d4b7b4b9948ff968c6e272a785b6dd7b)

**Comment by ipdae** [comment]
test is roken

**Review by Unengine — APPROVED** [review]

**Review by ipdae — APPROVED** [review]

**Review by boscohyun — APPROVED** [review]

File: Lib9c/Model/State/AvatarState.cs [hunk 5]
```diff
@@ -545,6 +595,33 @@ public void EquipCostumes(HashSet<int> costumeIds)
...
+            foreach (var costumeId in costumeIds)
+            {
+                if (!inventory.TryGetNonFungibleItem<Costume>(costumeId, out var costume))
+                {
+                    continue;
+                }
+
+                costume.Equip();
+            }
+        }
```
Comment (boscohyun) [inline Lib9c/Model/State/AvatarState.cs:624]:
```suggestion
```

https://github.com/planetarium/lib9c/blob/6eb98d42ccd4cc7bff86688af6bbb74e01c84d19/Lib9c/Model/State/AvatarState.cs#L522

File: Lib9c/Action/MimisbrunnrBattle.cs [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            if (avatarState.actionPoint < stageRow.CostAP)
+            {
+                throw new NotEnoughActionPointException(
+                    $"Aborted due to insufficient action point: " +
+                    $"{avatarState.actionPoint} < {stageRow.CostAP}"
+                );
+            }
+            avatarState.actionPoint -= stageRow.CostAP;
+            avatarState.EquipCostumes(costumes);
+            avatarState.EquipEquipments(equipments);
```
Comment (boscohyun) [inline Lib9c/Action/MimisbrunnrBattle.cs:171]:
```suggestion
            avatarState.EquipItems(new List<Guid>(costumes).AddRange(equipments));
```

https://github.com/planetarium/lib9c/blob/6eb98d42ccd4cc7bff86688af6bbb74e01c84d19/Lib9c/Model/State/AvatarState.cs#L522

**Review by boscohyun — CHANGES_REQUESTED** [review]

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+                RankingMapAddress = _rankingMapAddress,
+            };
+            nextState = hackAndSlash.Execute(new ActionContext
+            {
+                PreviousStates = nextState,
+                Signer = _agentAddress,
+                Random = new TestRandom(),
+                Rehearsal = false,
+            });
+            Assert.True(hackAndSlash.Result.IsClear);
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:192]: These lines seems unnecessary.

```suggestion
```

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+                {
+                    PreviousStates = nextState,
+                    Signer = _agentAddress,
+                    Random = new TestRandom(),
+                    Rehearsal = false,
+                });
+            });
+
+            avatarState = nextState.GetAvatarState(_avatarAddress);
+            Assert.True(avatarState.worldInformation.IsStageCleared(stageIdToClear));
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:220]: This line seems unnecessary.

```suggestion
```

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+            Assert.False(avatarState.worldInformation.IsWorldUnlocked(worldIdToUnlock));
+
+            var tableCsv = _initialState.GetSheetCsv<WorldUnlockSheet>();
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:223]:
```suggestion
            var tableCsv = nextState.GetSheetCsv<WorldUnlockSheet>();
```

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+                Random = new TestRandom(),
+                Rehearsal = false,
+            });
+            Assert.True(hackAndSlash.Result.IsClear);
+
+            avatarState = nextState.GetAvatarState(_avatarAddress);
+            Assert.True(avatarState.worldInformation.IsStageCleared(stageIdToClear));
+            Assert.False(avatarState.worldInformation.IsWorldUnlocked(worldIdToUnlock));
+
+            var tableCsv = _initialState.GetSheetCsv<WorldUnlockSheet>();
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:113]:
```suggestion
            var tableCsv = nextState.GetSheetCsv<WorldUnlockSheet>();
```

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+            var worldUnlockSheet = _initialState.GetSheet<WorldUnlockSheet>();
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:114]:
```suggestion
            var worldUnlockSheet = nextState.GetSheet<WorldUnlockSheet>();
```

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+                    Rehearsal = false,
+                });
+            });
+
+            avatarState = nextState.GetAvatarState(_avatarAddress);
+            Assert.True(avatarState.worldInformation.IsStageCleared(stageIdToClear));
+            Assert.False(avatarState.worldInformation.IsWorldUnlocked(worldIdToUnlock));
+
+            var tableCsv = _initialState.GetSheetCsv<WorldUnlockSheet>();
+            var newTable = new StringBuilder(tableCsv).Replace("4,2,100,10001", "4,2,80,10001").ToString();
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:224]: We need to subtract the hardcoded part of this line as a factor. For example..

```suggestion
            var newTable = new StringBuilder(tableCsv).Replace(targetRowStringBefore, targetRowStringAfter).ToString();
```

File: .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+            {
+                PreviousStates = nextState,
+                Signer = _agentAddress,
+                Random = new TestRandom(),
+                Rehearsal = false,
+            });
+
+            avatarState = nextState.GetAvatarState(_avatarAddress);
+            Assert.True(hackAndSlash.Result.IsClear);
+            Assert.True(avatarState.worldInformation.IsWorldUnlocked(worldIdToUnlock));
```
Comment (boscohyun) [inline .Lib9c.Tests/Action/Scenario/WorldUnlockScenarioTest.cs:253]: These lines seems unnecessary.

```suggestion
```

**Review by boscohyun — CHANGES_REQUESTED** [review]

**Review by Unengine — APPROVED** [review]

File: Lib9c/Battle/StageRewardExpHelper.cs [hunk 14]
```diff
@@ -45,6 +45,11 @@ public static class StageRewardExpHelper
 
         public static int GetExp(int characterLevel, int stageNumber)
         {
+            if (stageNumber >= GameConfig.MimisbrunnrWorldId)
```
[the remark this answers is not in the document]
Comment (hskim881028) [reply Lib9c/Battle/StageRewardExpHelper.cs]: It's a miracle to go well.. I should have compared the stageId

**Review by boscohyun — APPROVED** [review]

**Pull request merged**